ton_client/src/boc/tvc.rs

```diff
@@ -1,387 +1,123 @@
-    let code_hash = tvc.object.code.as_ref().map(|cell| cell.repr_hash().as_hex_string());
-    let compiler_version = tvc.object.code.clone()
-        .map(|cell| get_compiler_version_from_cell(cell).ok())
-        .flatten()
-        .flatten();
-    let code = serialize("code", tvc.object.code, params.boc_cache.clone()).await?;
-
-    let data_depth = tvc.object.data.as_ref().map(|cell| cell.repr_depth() as u32);
-    let data_hash = tvc.object.data.as_ref().map(|cell| cell.repr_hash().as_hex_string());
-    let data = serialize("data", tvc.object.data, params.boc_cache.clone()).await?;
-
-    let library = serialize("library", tvc.object.library.root().cloned(), params.boc_cache.clone()).await?;
-
-    Ok(ResultOfDecodeTvc {
-        code,
-        code_depth,
-        code_hash,
-        data,
-        data_depth,
-        data_hash,
-        library,
-        tick: tvc.object.special.as_ref().map(|val| val.tick),
-        tock: tvc.object.special.as_ref().map(|val| val.tick),
-        split_depth: tvc.object.split_depth.map(|val| val.as_u32()),
-        compiler_version,
-    })
-}
+/*
+    Copyright 2023 EverX Labs.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+use crate::boc::internal::{
+    deserialize_cell_from_boc, deserialize_object_from_boc, serialize_cell_to_base64,
+    serialize_object_to_cell,
+};
+use crate::boc::tvc_serialization::{TvmSmc, Version, TVC};
+use crate::error::ClientResult;
+use crate::ClientContext;
+use ton_block::StateInit;
+use ton_types::Cell;
+
+#[derive(Serialize, Deserialize, ApiType, Default)]
+pub struct ParamsOfDecodeTvc {
+    /// Contract TVC BOC encoded as base64 or BOC handle
+    pub tvc: String,
+}
+
+#[derive(Serialize, ApiType, Eq, PartialEq, Debug)]
+pub struct ResultOfDecodeTvc {
+    /// Decoded TVC
+    pub tvc: Tvc,
+}
+
+#[derive(Serialize, ApiType, Eq, PartialEq, Debug)]
+#[serde(tag = "type", content = "value")]
+pub enum Tvc {
+    None,
+    V1(TvcV1),
+}
+
+#[derive(Serialize, ApiType, Eq, PartialEq, Debug)]
+pub struct TvcV1 {
+    pub code: String,
+    pub meta: Option<TvcV1Metadata>,
+}
+
+#[derive(Serialize, ApiType, Eq, PartialEq, Debug)]
+pub struct TvcV1Metadata {
+    pub sold: TvcV1Version,
+    pub linker: TvcV1Version,
+    pub compiled_at: String,
+    pub name: String,
+    pub desc: String,
+}
+
+#[derive(Serialize, ApiType, Eq, PartialEq, Debug)]
+pub struct TvcV1Version {
```

Comment:
Better name it `TvcVersion` or `TvcVersionV1` since the same version struct may be used in several Tvc struct versions